streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/HighAvailabilityTaskAssignor.java

```diff
@@ -0,0 +1,548 @@
+    private final Map<ID, Integer> clientsToNumberOfThreads;
+    private final SortedSet<ID> sortedClients;
+
+    private final Set<TaskId> allTasks;
+    private final SortedSet<TaskId> statefulTasks;
+    private final SortedSet<TaskId> statelessTasks;
+
+    private final AssignmentConfigs configs;
+
+    private final SortedMap<TaskId, SortedSet<RankedClient<ID>>> statefulTasksToRankedCandidates;
+
+    public HighAvailabilityTaskAssignor(final Map<ID, ClientState> clientStates,
+                                        final Set<TaskId> allTasks,
+                                        final Set<TaskId> statefulTasks,
+                                        final AssignmentConfigs configs) {
+        this.configs = configs;
+        this.clientStates = clientStates;
+        this.allTasks = allTasks;
+        this.statefulTasks = new TreeSet<>(statefulTasks);
+
+        statelessTasks = new TreeSet<>(allTasks);
+        statelessTasks.removeAll(statefulTasks);
+
+        sortedClients = new TreeSet<>();
+        clientsToNumberOfThreads = new HashMap<>();
+        clientStates.forEach((client, state) -> {
+            sortedClients.add(client);
+            clientsToNumberOfThreads.put(client, state.capacity());
+        });
+
+        statefulTasksToRankedCandidates =
+            buildClientRankingsByTask(statefulTasks, clientStates, configs.acceptableRecoveryLag);
+    }
+
+    @Override
+    public boolean assign() {
+        if (shouldUsePreviousAssignment()) {
+            assignPreviousTasksToClientStates();
+            return false;
+        }
+
+        final Map<ID, List<TaskId>> warmupTaskAssignment = initializeEmptyTaskAssignmentMap();
+        final Map<ID, List<TaskId>> standbyTaskAssignment = initializeEmptyTaskAssignmentMap();
+        final Map<ID, List<TaskId>> statelessActiveTaskAssignment = initializeEmptyTaskAssignmentMap();
+
+        // ---------------- Stateful Active Tasks ---------------- //
+
+        final Map<ID, List<TaskId>> statefulActiveTaskAssignment =
+            new DefaultStateConstrainedBalancedAssignor<ID>().assign(
+                statefulTasksToRankedCandidates,
+                configs.balanceFactor,
+                sortedClients,
+                clientsToNumberOfThreads
+            );
+
+        // ---------------- Warmup Replica Tasks ---------------- //
+
+        final Map<ID, List<TaskId>> balancedStatefulActiveTaskAssignment =
+            new DefaultBalancedAssignor<ID>().assign(
+                sortedClients,
+                statefulTasks,
+                clientsToNumberOfThreads,
+                configs.balanceFactor);
+
+        final List<Movement<ID>> movements =
+            getMovements(statefulActiveTaskAssignment, balancedStatefulActiveTaskAssignment,
+                configs.maxWarmupReplicas);
+        for (final Movement<ID> movement : movements) {
+            warmupTaskAssignment.get(movement.destination).add(movement.task);
+        }
```

Comment:
Q: IIUC, we do not check if the movement is for free. That is, if the destination is a caught-up client. If it were we would not need to assign a warm-up replica and could consider one more movement. I am also fine with post-poning that to a follow-up PR.

Reply:
Oh, that's a really good point. We should totally do that, but I agree it would be easier to do it in a follow-on.

Reply:
Does this cover what you're proposing (minus the optimizations) https://confluentinc.atlassian.net/browse/KSTREAMS-3938

Reply:
Sounds good